events/eventservices/eventService.go

```diff
@@ -1,13 +1,266 @@
 package eventservices
 
-import "github.com/FactomProject/factomd/events"
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/FactomProject/factomd/common/constants/runstate"
+	"github.com/FactomProject/factomd/common/globals"
+	"github.com/FactomProject/factomd/events"
+	"github.com/FactomProject/factomd/events/eventinput"
+	"github.com/FactomProject/factomd/events/eventmessages/generated/eventmessages"
+	"github.com/FactomProject/factomd/events/events_config"
+	"github.com/FactomProject/factomd/p2p"
+	"github.com/FactomProject/factomd/util"
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	"net"
+	"reflect"
+	"time"
+)
 
-type EventService interface {
-	Send(event events.EventInput) error
+var eventService events.EventService
+var eventServiceControl events.EventServiceControl
+
+const (
+	defaultProtocol       = "tcp"
+	defaultConnectionHost = "127.0.0.1"
+	defaultConnectionPort = 8040
+	defaultOutputFormat   = events_config.Protobuf
+	protocolVersion       = byte(1)
+)
+
+var (
+	dialRetryPostponeDuration = 5 * time.Minute
+	redialSleepDuration       = 10 * time.Second
+	sendRetries               = 3
+)
+
+type eventServiceInstance struct {
+	params                  *EventServiceParams
+	eventsOutQueue          chan *eventmessages.FactomEvent
+	postponeSendingUntil    time.Time
+	connection              net.Conn
+	parentState             events.StateEventServices
+	droppedFromQueueCounter prometheus.Counter
+	notSentCounter          prometheus.Counter
+}
+
+func NewEventService(state events.StateEventServices, config *util.FactomdConfig, factomParams *globals.FactomParams) events.EventServiceControl {
+	return NewEventServiceTo(state, selectParameters(factomParams, config))
+}
+
+func NewEventServiceTo(state events.StateEventServices, params *EventServiceParams) events.EventServiceControl {
+	if eventService == nil {
+		eventServiceInstance := &eventServiceInstance{
+			eventsOutQueue: make(chan *eventmessages.FactomEvent, p2p.StandardChannelSize),
+			params:         params,
+			parentState:    state,
+		}
+		eventService = eventServiceInstance
+		eventServiceControl = eventServiceInstance
+		events.AttachEventServiceToEventEmitter(state, eventServiceInstance)
+
+		eventServiceInstance.droppedFromQueueCounter = prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "factomd_livefeed_dropped_from_queue_counter",
+			Help: "Number of times we dropped events due of a full the event queue",
+		})
+		eventServiceInstance.notSentCounter = prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "factomd_livefeed_not_send_counter",
+			Help: "Number of times we couldn't send out an event",
+		})
+
+		go eventServiceInstance.processEventsChannel()
+	}
+	return eventServiceControl
```

Comment:
return eventServiceControl after null check of eventService.

Reply:
yes, when it's null it creates an eventServiceInstance == eventServiceControl otherwise it returns the existing one